src/Database/LSMTree/Internal/Vector.hs

```diff
@@ -95,3 +96,24 @@ unsafeInsertWithMStrict ::
 unsafeInsertWithMStrict mvec f i y = VM.unsafeModifyM mvec g i
   where
     g x = pure $! Just $! maybe y (`f` y) x
+
+{-# INLINE unfoldrNM' #-}
+-- | A version of 'V.unfoldrNM' that also returns the final state.
+--
+-- /O(n)/ Construct a vector by repeatedly applying the monadic generator
+-- function to a seed. The generator function also yields 'Just' the next
+-- element or 'Nothing' if there are no more elements.
+unfoldrNM' :: PrimMonad m => Int -> (b -> m (Maybe a, b)) -> b -> m (V.Vector a, b)
+unfoldrNM' len f = \b0 -> do
+    vec <- VM.unsafeNew len
+    go vec 0 b0
+  where
+    go vec !n !b
+      | n >= len = (, b) <$> V.unsafeFreeze vec
+      | otherwise =
+          f b >>= \case
+            (Nothing, b') ->
+              (, b') <$> V.unsafeFreeze (VM.slice 0 n vec)
+            (Just a,  b') -> do
+              VM.unsafeWrite vec n a
```

Comment:
Also, you might make this element strict, i.e. use `unsafeWrite vec n $! a`. This could also be reflected in the name -- and it might deserve a comment on just how strict the result actually is

Reply:
I did it by adding a bang on the pattern match on `a` and added a comment.